File: solutions/erlang/robot-simulator/1/src/robot_simulator.erl
-module(robot_simulator).

-export([advance/1, create/0, direction/1, left/1, place/3, position/1, right/1,
         robot_state/1]).

advance(Robot) ->
    Robot ! advance.

create() ->
    Id = spawn(robot_simulator, robot_state, [#{"position" => {0, 0}, "direction" => north}]),
    Id.

direction(Robot) ->
    Robot ! {direction, self()},
    receive
        V ->
            V
    end.

left(Robot) ->
    Robot ! left.

place(Robot, Direction, Position) ->
    Robot ! {place, #{"position" => Position, "direction" => Direction}}.

position(Robot) ->
    Robot ! {position, self()},
    receive
        V ->
            V
    end.

right(Robot) ->
    Robot ! right.

robot_state(Robot_state) ->
    receive
        left ->
            Dir = maps:get("direction", Robot_state),
            case Dir of
                north ->
                    robot_state(maps:update("direction", west, Robot_state));
                east ->
                    robot_state(maps:update("direction", north, Robot_state));
                south ->
                    robot_state(maps:update("direction", east, Robot_state));
                west ->
                    robot_state(maps:update("direction", south, Robot_state))
            end;
        right ->
            Dir = maps:get("direction", Robot_state),
            case Dir of
                north ->
                    robot_state(maps:update("direction", east, Robot_state));
                east ->
                    robot_state(maps:update("direction", south, Robot_state));
                south ->
                    robot_state(maps:update("direction", west, Robot_state));
                west ->
                    robot_state(maps:update("direction", north, Robot_state))
            end;
        advance ->
            Position = maps:get("position", Robot_state),
            Direction = maps:get("direction", Robot_state),
            case Direction of
                north ->
                    robot_state(maps:update("position",
                                            {element(1, Position), element(2, Position) + 1},
                                            Robot_state));
                east ->
                    robot_state(maps:update("position",
                                            {element(1, Position) + 1, element(2, Position)},
                                            Robot_state));
                south ->
                    robot_state(maps:update("position",
                                            {element(1, Position), element(2, Position) - 1},
                                            Robot_state));
                west ->
                    robot_state(maps:update("position",
                                            {element(1, Position) - 1, element(2, Position)},
                                            Robot_state))
            end;
        {place, Robot_state_new} ->
            robot_state(Robot_state_new);
        {direction, From} ->
            From ! maps:get("direction", Robot_state);
        {position, From} ->
            From ! maps:get("position", Robot_state)
    end,
    robot_state(Robot_state).

File: solutions/elixir/robot-simulator/1/lib/robot_simulator.ex
defmodule RobotSimulator do
  @directions %{:north => 0, :east => 1, :south => 2, :west => 3}
  defstruct position: {0, 0}, direction: :north

  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction :: atom, position :: {integer, integer}) :: %RobotSimulator{}
  def create(direction \\ :north, position \\ {0, 0}) do
    cond do
      not Map.has_key?(@directions, direction) ->
        {:error, "invalid direction"}

      not is_tuple(position) or tuple_size(position) != 2 or
          not (fn {a, b} -> is_integer(a) and is_integer(b) end).(position) ->
        {:error, "invalid position"}

      true ->
        %RobotSimulator{direction: direction, position: position}
    end
  end

  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: any, instructions :: String.t()) :: any
  def simulate(robot, instructions) do
    simulate_helper(robot, to_charlist(instructions))
  end

  defp simulate_helper(robot, []), do: robot

  defp simulate_helper(robot, [head | tail]) do
    case head do
      ?R ->
        simulate_helper(
          %{
            robot
            | direction: enum_map_helper(@directions, rem(@directions[robot.direction] + 1, 4))
          },
          tail
        )

      ?L ->
        simulate_helper(
          %{
            robot
            | direction:
                enum_map_helper(@directions, rem(4 + @directions[robot.direction] - 1, 4))
          },
          tail
        )

      ?A ->
        simulate_helper(advance_helper(robot), tail)

      _ ->
        {:error, "invalid instruction"}
    end
  end

  defp enum_map_helper(map, val) do
    {key, _} = Enum.find(map, fn {_, value} -> val == value end)
    key
  end

  defp advance_helper(robot) do
    {posx, posy} = robot.position

    case robot.direction do
      :north ->
        %{robot | position: {posx, posy + 1}}

      :east ->
        %{robot | position: {posx + 1, posy}}

      :south ->
        %{robot | position: {posx, posy - 1}}

      :west ->
        %{robot | position: {posx - 1, posy}}
    end
  end

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: %RobotSimulator{}) :: atom
  def direction(robot) do
    robot.direction
  end

  @doc """
  Return the robot's position.
  """
  @spec position(robot :: %RobotSimulator{}) :: {integer, integer}
  def position(robot) do
    robot.position
  end
end
